Fix failfast stacktrace

sywhang opened the pull request:
This addresses issue #14867, by adding another Environment.FailFast FCall that gets passed in an UINT type that specifies the error source, to print the correct title for stack trace.

The stack getting printed twice is also addressed here.

noahfalk commented on src/mscorlib/shared/System/Diagnostics/Debug.cs:
```diff
@@ -13,6 +13,7 @@ namespace System.Diagnostics
     public static partial class Debug
     {
         private static readonly object s_lock = new object();
+        private enum ErrorSource { Assert=1, Assume }
```
You should add comments both here and in eepolicy.h indicating that these enumerations mirror one another and must be kept in sync. I would also name this one identical to what is eepolicy.

jkotas replied:
Can we just pass this as a string instead of enum?

sywhang replied:
My initial implementation was passing around string, but I thought it was easier to pass around an enum. I'll fix it to string. Thanks!

jkotas commented on src/mscorlib/src/System/Diagnostics/Contracts/ContractsBCL.cs:
```diff
@@ -318,7 +318,7 @@ public static void TriggerFailure(ContractFailureKind kind, String displayMessag
             }
 
             // TODO: https://github.com/dotnet/coreclr/issues/14867
```
You should delete the TODO

jkotas commented on src/mscorlib/shared/System/Diagnostics/Debug.cs:
```diff
@@ -117,6 +135,18 @@ public static void Fail(string message, string detailMessage)
             Assert(false, message, detailMessage);
         }
+        
+        private static string PopStackTrace(string stackTrace)
```
Instead of this method, use the StackTrace constructor that lets you specify number of methods to ship. (Take a look at the implementation of EnvironmentAugments.StackTrace.) It will avoid need to reconstruct the stacktrace.

jkotas commented on src/vm/eepolicy.cpp:
```diff
@@ -1176,18 +1176,31 @@ inline void LogCallstackForLogWorker()
         PrintToStdErrA("\n");
 
-        if (pThread)
+        if (pThread && errorSource == EEPolicy::FFES_FailFast)
```
This looks like that we will print stack trace for assertion failed, and then the text of the error. It would be nice always have the error first - it is how it always worked.

noahfalk commented on src/mscorlib/shared/System/Diagnostics/Debug.cs:
```diff
@@ -91,18 +92,35 @@ public static void Assert(bool condition, string message, string detailMessage)
+        }
 
+        [System.Diagnostics.Conditional("DEBUG")]
+        public static void Assume(bool condition, string message, string detailMessage)
```
Lets make this internal. I don't think there was any intent to add new public APIs and if we want to add it later we can always change it then.

noahfalk replied:
The naming here appears to favor 'Assume' but there are a number of other kinds of contract method such as Invariant or Requires. I'd suggest doing ContractFail(string message, string detailMessage). I would correspondingly change the error text that gets printed to "Contract Failed" unless @jkotas was particular that it really be "Assumption failed"?

jkotas replied:
> Assumption failed

It is what it prints on .NET Framework. I am not particular about the name. ContrailFail sounds fine to me. BTW: A related method is called `OnContractFailure` in CoreRT/ProjectN.

noahfalk replied:
Ah if that is what .NET framework prints then I'd gladly make the behavior match. I thought we had inherited this bad behavior from .Net Framework however. Do you happen to know if something specifically regressed the .Net Core behavior? If you don't know offhand no worries, Sung and I can compare them.

Thanks Jan!

jkotas replied:
The specific change that touched the `Assumption failed` message was my #14868. Contracts used its own path to display error message with its own set of issues. When I have switched them to use the same path as asserts, I have found that asserts are broken in multiple different ways so I just left TODO there.

Other changes that contributed to this:
- Original refactoring of `System.Diagnostic.Debug` namespace for CoreCLR and ProjectN
- Assert was changed to now show message box by default in .NET Core (message boxes do not make sense in cloud environments)
- FailFast was changed to print the stacktrace in .NET Core (it did nothing in .NET Framework). This change caused stacktrace to be printed twice for Assert.

noahfalk commented on src/vm/eepolicy.cpp:
```diff
@@ -1176,18 +1176,31 @@ inline void LogCallstackForLogWorker()
+        {
+            PrintToStdErrA("FailFast: ");
+        }
+        else if (errorSource == EEPolicy::FFES_FailFast)
```
I don't think this works : ) All the clauses are the same.

sywhang replied:
Oops, thanks for spotting that :)

jkotas commented on src/mscorlib/shared/System/Diagnostics/Debug.cs:
```diff
@@ -315,7 +345,8 @@ private sealed class DebugAssertException : Exception
 
         // internal and not readonly so that the tests can swap this out.
-        internal static Action<string, string, string> s_ShowAssertDialog = ShowAssertDialog;
+        internal static Action<string, string, string, uint> s_ShowDialog = ShowDialog;
```
This file is in the shared CoreLib partition. It is shared between CoreCLR and CoreRT/ProjectN. There will be matching change needed in ProjectN.

sywhang replied:
I'll talk to Noah about making a matching change in ProjectN. Thanks for letting me know!

danmosemsft commented:
cc @ViktorHofer

jkotas commented on src/classlibnative/bcltype/system.cpp:
```diff
@@ -423,6 +423,15 @@ void SystemNative::GenericFailFast(STRINGREF refMesgString, EXCEPTIONREF refExce
+
+    if (errorSource != NULL) {
+        DWORD cchErrorSource = errorSource->GetStringLength();
+        errorSourceString = new (nothrow) WCHAR[cchErrorSource + 1];
```
`(nothrow)` means that this can return null. You should probably just ignore the passed in string when that happens.

ViktorHofer commented on src/vm/eepolicy.cpp:
```diff
@@ -1176,18 +1176,31 @@ inline void LogCallstackForLogWorker()
-        PrintToStdErrA("FailFast: ");
+        if (errorSource == NULL)
+        {
+            PrintToStdErrA("FailFast: ");
```
The failfast message will now be printed in the next line. Intentionally? If yes, you could deliver the white space character

ViktorHofer replied:
Remove*

danmosemsft commented:
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test 
@dotnet-bot test Ubuntu arm Cross Debug Innerloop Build

sywhang commented:
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Debug Innerloop Build

sywhang commented:
@jkotas @noahfalk @ViktorHofer, Thank you all for the feedback! I believe I addressed all the comments. Could you please take another look at this PR? Thanks :)

jkotas commented on src/mscorlib/shared/System/Diagnostics/Debug.cs:
```diff
@@ -91,18 +91,34 @@ public static void Assert(bool condition, string message, string detailMessage)
+                }
                 WriteLine(FormatAssert(stackTrace, message, detailMessage));
-                s_ShowAssertDialog(stackTrace, message, detailMessage);
+                s_ShowDialog(stackTrace, message, detailMessage, "Assumption Failed");
```
This may rather call `SR.GetResourceString(GetResourceNameForFailure(failureKind))` instead of hard-coding `"Assumption Failed"`.

jkotas commented on src/mscorlib/shared/System/Diagnostics/Debug.cs:
```diff
@@ -91,18 +91,34 @@ public static void Assert(bool condition, string message, string detailMessage)
+            }
+        }
 
+        internal static void Assume(bool condition, string message, string detailMessage)
```
This method is called from multiple contract methods, not just Assume, e.g. it is called from `Contract.Assert`. `ContractFailure` maybe a better name for this method.

noahfalk commented on src/classlibnative/bcltype/system.cpp:
```diff
@@ -377,7 +377,7 @@ WCHAR g_szFailFastBuffer[256];
 // This is the common code for FailFast processing that is wrapped by the two
 // FailFast FCalls below.
-void SystemNative::GenericFailFast(STRINGREF refMesgString, EXCEPTIONREF refExceptionForWatsonBucketing, UINT_PTR retAddress, UINT exitCode)
+void SystemNative::GenericFailFast(STRINGREF refMesgString, EXCEPTIONREF refExceptionForWatsonBucketing, UINT_PTR retAddress, UINT exitCode, STRINGREF errorSource)
```
When managed code passes managed object references to native code we have to take extra precautions to ensure we don't create GC holes. Check out: https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md#2.1
In this case you are using errorSource unsafely.

noahfalk commented on src/classlibnative/bcltype/system.cpp:
```diff
@@ -539,7 +553,27 @@ FCIMPL2(VOID, SystemNative::FailFastWithException, StringObject* refMessageUNSAF
+    EXCEPTIONREF refException = (EXCEPTIONREF)refExceptionUNSAFE;
+    STRINGREF errorSource = (STRINGREF)errorSourceUNSAFE;
+
+    HELPER_METHOD_FRAME_BEGIN_2(refMessage, refException);
```
Another GC hole here.

jkotas commented:
test Ubuntu x64 Checked corefx_baseline please
test Windows_NT x64 Checked corefx_baseline please

sywhang commented:
test Ubuntu x64 Checked corefx_baseline please
test Windows_NT x64 Checked corefx_baseline please
